### PR TITLE
fix(components/forms)!: don't show errors till blur

### PR DESCRIPTION
**Pull request by Blackbaud-SandhyaRajasabeson**
[AB#2995433](https://dev.azure.com/blackbaud/Products/_boards/board/t/SKY%20UX%20Program/Stories/?workitem=2995433)

### CODE REVIEW

File: libs/components/forms/src/lib/modules/input-box/input-box.component.html [hunk 1]
```diff
@@ -184,23 +181,24 @@
     [errors]="controlDir?.errors"
     [labelText]="labelText"
     [ngClass]="{ 'sky-screen-reader-only': errorsScreenReaderOnly() }"
-    [showErrors]="controlDir?.touched || controlDir?.dirty"
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/input-box/input-box.component.html:187]: Most of the changes in this PR are linting issues, I'm going to mark the relevant changes

File: libs/components/forms/src/lib/modules/input-box/input-box.component.spec.ts [hunk 2]
```diff
@@ -344,7 +344,7 @@ describe('Input box component', () => {
 
     fixture.detectChanges();
 
-    validateInvalid('when dirty and untouched', inputBoxEl, true);
+    validateInvalid('when dirty and untouched', inputBoxEl, false);
   }
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/input-box/input-box.component.spec.ts:348]: relevant change

File: libs/components/forms/src/lib/modules/input-box/input-box.component.ts [hunk 3]
```diff
@@ -339,7 +339,7 @@ export class SkyInputBoxComponent
   }
 
   #controlHasErrors(control: AbstractControlDirective | undefined): boolean {
-    return !!(control && control.invalid && (control.dirty || control.touched));
+    return !!(control && control.invalid && control.touched);
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/input-box/input-box.component.ts:342]: relevant change

File: libs/components/forms/src/lib/modules/input-box/input-box.component.html (hunk 1 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/forms/src/lib/modules/input-box/input-box.component.html:187]: Checking for either "dirty" OR "touched" is an Angular-prescribed condition for when to show error messages. The issue is more than likely with our form components and when we decide to mark a control as "dirty". According to Angular, a form control should be marked as "dirty" when the value has changed. Did you dive into the affected components to see if they're marking the control as "dirty" improperly?

> When the user changes the value in the watched field, the control is marked as "dirty"

See: https://v17.angular.io/guide/form-validation

**Comment by blackbaud-sky-build-user** [comment]
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/2602/storybook/)

Component Storybooks:

- [action-bars-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/action-bars-storybook/)
- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/ag-grid-storybook/)
- [angular-tree-component-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/angular-tree-component-storybook/)
- [avatar-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/avatar-storybook/)
- [colorpicker-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/colorpicker-storybook/)
- [data-manager-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/data-manager-storybook/)
- [datetime-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/datetime-storybook/)
- [errors-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/errors-storybook/)
- [forms-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/forms-storybook/)
- [indicators-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/indicators-storybook/)
- [inline-form-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/inline-form-storybook/)
- [layout-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/layout-storybook/)
- [lists-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/lists-storybook/)
- [lookup-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/lookup-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/pages-storybook/)
- [phone-field-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/phone-field-storybook/)
- [split-view-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/split-view-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/tabs-storybook/)
- [text-editor-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/text-editor-storybook/)
- [theme-storybook](https://blackbaud.github.io/skyux-pr-preview/2602/storybooks/theme-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/2602/code-examples/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/2602/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/2602/playground/)

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.html [hunk 4]
```diff
@@ -1,104 +1,133 @@
-<ng-container *ngIf="labelText && showErrors">
-  <ng-container *ngIf="errors">
-    <sky-form-error
-      *ngIf="errors['required']"
-      errorName="required"
-      [errorText]="'skyux_form_error_required' | skyLibResources: labelText"
-    />
+@if (labelText && showErrors) {
+  <ng-container>
```
Comment (Blackbaud-CoreyArcher) [inline libs/components/forms/src/lib/modules/form-error/form-errors.component.html]: These `ng-container`s are no longer needed. Containers don't exist in the DOM, they purely exist to attach structural directives to. In fact, when you put a structural directive on a real element, it actually gets wrapped in a container that is showing/hiding or replicating or whatever. The new control flow syntax just makes what's actually happening with the DOM manipulation more apparent.

File: libs/components/forms/src/lib/modules/input-box/input-box.component.html (hunk 1 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/forms/src/lib/modules/input-box/input-box.component.html:187]: after confirming with design here are how our first class errors are meant to react after this story:

required: touched
maxlength & file attachment errors: dirty
all thers: touched && dirty

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html [hunk 5]
```diff
@@ -38,20 +38,18 @@
           }}
         </option>
       </select>
-    </sky-input-box>
-    <sky-form-errors [labelText]="labelText">
       <sky-form-error
         *ngIf="
-          hostControl?.errors?.['skyDateRange']?.errors.endDateBeforeStartDate
+          hostControl?.errors?.['skyDateRange']?.errors
+            .endDateBeforeStartDate && labelText
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.html:44]: date range picker had `sky-form-errors` outside of the dropdown input box to have the error span the entire length of date range picker. We have moved it into the first input box and the error will now show under the first box in date ranger picker. This is to keep it in alignment with all the other errors.

File: libs/components/forms/src/lib/modules/input-box/input-box.component.html [hunk 6]
```diff
@@ -171,36 +169,37 @@
...
-      *ngIf="hostIconsInsetLeftTemplate"
-      [ngTemplateOutlet]="hostIconsInsetLeftTemplate"
-    />
+    @if (hostIconsInsetLeftTemplate) {
+      <ng-container [ngTemplateOutlet]="hostIconsInsetLeftTemplate" />
+    }
   </div>
 </ng-template>
 
 <ng-template #errorLabelTemplate>
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/input-box/input-box.component.html:178]: the below chunk of code adding `sky-input-box-errors` is the only relevant change in this file.

File: libs/components/forms/src/lib/modules/input-box/input-box.component.html [hunk 7]
```diff
@@ -76,92 +76,90 @@
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/input-box/input-box.component.html:76]: the below chunk of code adding `sky-input-box-errors` on line 182/179 is the only relevant change in this file.

File: libs/components/forms/src/lib/modules/form-error/form-errors.component.ts [hunk 8]
```diff
@@ -30,12 +29,6 @@ import { SkyFormErrorComponent } from './form-error.component';
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class SkyFormErrorsComponent {
-  /**
-   * The validation errors from the form control.
-   */
-  @Input()
-  public errors: ValidationErrors | null | undefined;
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/forms/src/lib/modules/form-error/form-errors.component.ts:37]: `sky-form-errors` will no longer be handling any errors so this input is no longer needed

File: libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts [hunk 9]
```diff
@@ -430,7 +430,11 @@ export class SkyDateRangePickerComponent
         this.#changeDetector.markForCheck();
       }
 
-      this.hasErrors = !!control.errors && (control.touched || control.dirty);
+      if (this.#startDateControl.dirty || this.#endDateControl.dirty) {
+        this.#calculatorIdControl.markAsDirty();
+      }
```
Comment (Blackbaud-SandhyaRajasabeson) [inline libs/components/datetime/src/lib/modules/date-range-picker/date-range-picker.component.ts:435]: if either of the two date input boxes in date range picker are dirty, make sure the first input box (the dropdown) is also marked as dirty. That first dropdown is where all errors are shown for now.

**Comment by Blackbaud-SandhyaRajasabeson** [comment]
Closing in favor of this PR: https://github.com/blackbaud/skyux/pull/2624

- From a triad discussion a more simplified approach was decided that does not involve moving the errors out of form errors
- we want to keep this branch for future consideration of moving form errors out to individual components in the future

**Pull request closed without merging**